Extend r!giverole for self-assignable roles, add r!rmrole and r!roles commands.

garrett16r opened the pull request:
r!giverole was extended so that any users with the Member role can freely assign themselves seven different color roles (red, blue, green, purple, yellow, orange, pink). The command will check to make sure that the user doesn't have any other color roles already assigned before allowing them to self-assign one, and will refer them to r!rmrole if they do have one already. If the user enters a nonexistent role name, they will be referred to r!roles to view the available options.

r!roles is a simple set of messages to display which colors they can choose from, but this can certainly be updated later on to also use embeds and other nicer formatting tools.

With the extension of r!giverole, there is an easy format to follow in the code that will allow for future additions such as region roles, among other possibilities.

Once this change is merged and deployed, roles for each of the seven colors will need to be added, each starting with a capital letter in order to follow the format that the code expects. Putting these roles above the Member role in the list should allow them to take priority for name color.

Commit:
Added ability to self-assign color roles

+ Extended the r!giverole command to allow for color roles.
+ Checks for various cases, including users trying to request multiple color roles, nonexistant roles, etc.
+ Added r!roles command for users to view all available color roles (Could be updated for nicer appearance).

Commit:
Revert "Added ability to self-assign color roles"

This reverts commit a165953ac9b1d5c0aed2807f0210847d5b5195d7.

Commit:
Revert "Revert "Added ability to self-assign color roles""

This reverts commit 6249d395482e9d152e5cb0d8b4f737c9c57b6f41.

Commit:
Added r!rmrole command

Allows the user to remove any self-assigned role from themselves, other than the required Member role.

Commit:
Update rmrole.js

+ Fixed logical order of code to prevent incorrect responses to invalid input.
+ Improved responses to invalid input to be more informative.

Commit:
Removed all testing-related changes

- Reset constants to original values for r/Reol discord
- Removed start.bat
- Enabled all Redis modules

Commit:
Streamlined logical checks in giverole.js

Moved colors variable to COLORS in constants.js

Commit:
Update Readme.md

Commit:
Update roles.js

Cleaned up code showing the message that is send by r!roles

Commit:
Merge branch 'color-roles-dev' of https://github.com/Syntthetix/reol-chan into syn-color-roles-dev

0xbbadbeef commented on lib/commands/misc/giverole.js:
```diff
@@ -26,14 +26,38 @@ module.exports = class GiveRoleCommand extends Command {
     if (args.role === "member") {
-      const role = msg.guild.roles.find(r => r.name == 'Member')
-      if (!msg.member.roles.has(role)) {
+      const role = msg.guild.roles.find(r => r.name == 'Member');
```
It's best practice to always check with `===` because only a double `==` is a loose "abstract" comparison and may not give you the result you expect in some cases. Here's an [MDN of the difference](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness)

0xbbadbeef commented on lib/commands/misc/giverole.js:
```diff
@@ -26,14 +26,38 @@ module.exports = class GiveRoleCommand extends Command {
+				const memberRole = msg.guild.roles.find(r => r.name == 'Member');
+
+				// Confirms that the role provided by the user exists. If it doesn't, the method returns after informing the user.
+				try {
```
Let's not use try catch blocks as they're essentially "hacks" and 99% of the time there's a better way of catching an exception or catching negative scenarios. I'm not sure what data type `msg.guild.roles` is so I'm assuming it's a native `Map` class, it should just return `false` if the role was not found not throw an error which it seems you're expecting here.

We can simply refactor this verification like so

```
const roleExists = msg.guild.roles.has(role.id);
if (!roleExists) {
    return msg.reply('Invalid role! Available options can be found with r!roles');
}
```

causztic commented on lib/commands/misc/giverole.js:
```diff
@@ -26,14 +26,38 @@ module.exports = class GiveRoleCommand extends Command {
+			// Prevents this generalized code from allowing random people to self-assign the admin, moderator, or bae roles.
+    } else if (COLORS.includes(args.role)) {
+
+				const channel = this.client.channels.get(getChannel(REOL_CHAN_ID));
```
I think we can just set the code to be specifically runnable on the `bot stuff` channel, then we do not need a member check.

0xbbadbeef commented on lib/commands/misc/giverole.js:
```diff
@@ -26,14 +26,38 @@ module.exports = class GiveRoleCommand extends Command {
+				try {
+					let roleExists = msg.guild.roles.has(role.id);
+				} catch (err) {
+					msg.reply(`Invalid role! Available options can be found with r!roles`);
```
minor: Since we're not using string template syntax, the backticks are redundant. Please use single quotes or double quotes for regular strings where appropriate

causztic commented on lib/commands/misc/rmrole.js:
```diff
@@ -0,0 +1,53 @@
+	};
+
+  async run(msg, args) {
+    const channel = this.client.channels.get(getChannel(REOL_CHAN_ID));
```
i think we can change it to use the `bot stuff` channel instead - planning to rename the reol-chan channel just for new users as a landing page

causztic commented on lib/commands/misc/rmrole.js:
```diff
@@ -0,0 +1,53 @@
+    const memberRole = msg.guild.roles.find(r => r.name == 'Member');
+
+    // Prevent removal of required roles (Can be updated in the future if more required roles are ever added).
+    if (args.role === "Member" || args.role === "member") {
```
the role is called `Member`, so I don't think we need the second option. Since you already have `const role` above, you could use that instead of `args.role`.. maybe something like `role === memberRole`?

0xbbadbeef replied:
either way, it's probably also best lower case the role and the string literal we're comparing to. e.g. `args.role.toLowerCase() === 'member'`

0xbbadbeef commented on lib/commands/misc/rmrole.js:
```diff
@@ -0,0 +1,53 @@
+
+  async run(msg, args) {
+    const channel = this.client.channels.get(getChannel(REOL_CHAN_ID));
+    const role = msg.guild.roles.find(r => r.name == args.role);
```
I feel like we should be checking if we have a value for `args.role` first, just in case the user has only issued the command with no arguments. (although this could be handled internally, I'm not sure)

0xbbadbeef commented on lib/commands/misc/rmrole.js:
```diff
@@ -0,0 +1,53 @@
+    const memberRole = msg.guild.roles.find(r => r.name == 'Member');
+
+    // Prevent removal of required roles (Can be updated in the future if more required roles are ever added).
+    if (args.role === "Member" || args.role === "member") {
```
minor: to make things a little neater we can assign `args.role` into a const at the top of this function instead of doing `args.role` everywhere

0xbbadbeef commented on lib/commands/misc/rmrole.js:
```diff
@@ -0,0 +1,53 @@
+    if (args.role === "Member" || args.role === "member") {
+      msg.reply("You can't remove that role from yourself!");
+      return;
+    // Remove the role requested if the user is already a Member and has the provided role
```
I believe this comment has been forgotten to be removed?

causztic commented on lib/commands/misc/rmrole.js:
```diff
@@ -0,0 +1,53 @@
+      return;
+    }
+
+    if (msg.member.roles.has(memberRole.id) && msg.member.roles.has(role.id)) {
```
I don't think we need to check for the `memberRole` here because non-members will not have any other roles anyway.

0xbbadbeef replied:
minor: to clean up this condition, we can also assign `msg.member.roles` to a const before reading from it twice here.

e.g.
```
const memberRoles = msg.member.roles;
if (memberRoles.has(memberRole.id) && memberRoles.has(role.id)) {
```

causztic commented on lib/constants.js:
```diff
@@ -16,4 +17,4 @@ module.exports = {
   DISCOGRAPHY_ID
-};
\ No newline at end of file
+};
```
preserve the new line please!

causztic commented on lib/index.js:
```diff
@@ -106,4 +106,4 @@ client.on("message", message => {
   }
-})
\ No newline at end of file
+})
```
preserve the new line please!

causztic commented on lib/redis-client.js:
```diff
@@ -1,3 +1,3 @@
 const redisClient = redis.createClient(process.env.REDISTOGO_URL);
-module.exports = redisClient;
\ No newline at end of file
+module.exports = redisClient;
```
preserve the new line please!

0xbbadbeef commented on lib/constants.js:
```diff
@@ -7,6 +7,7 @@ REOL_CHAN_ID  = '296232884623376384'
 ANNOUNCEMENTS_ID = '413504832356352020'
 PHOTO_GALLERY_ID = '423787873448493056'
 DISCOGRAPHY_ID = '515055814755614730'
+COLORS = ["Red", "Blue", "Green", "Purple", "Yellow", "Orange", "Pink"]
```
it might be fun to not have hardcoded colours and allow users to assign their own hex for their custom role, to spice things up a bit. That could be your next task 😄

causztic replied:
@0xbbadbeef the roles we have on the server are region-based instead of colors, so that we can easily see who is from which region

0xbbadbeef replied:
@causztic ahh, okie

causztic replied:
@0xbbadbeef ok that'll be another PR, but yeah colors are tied to roles so for hexes for custom role will not be possible with the current discord API because that would require new colors being created on the fly..and there are limited roles I can create on the server.

0xbbadbeef commented on lib/commands/misc/giverole.js:
```diff
@@ -26,14 +26,38 @@ module.exports = class GiveRoleCommand extends Command {
+
+				const channel = this.client.channels.get(getChannel(REOL_CHAN_ID));
+				const role = msg.guild.roles.find(r => r.name == args.role);
+				const memberRole = msg.guild.roles.find(r => r.name == 'Member');
```
It might be best to set up a helper of some sorts that gets these 3 values as we're using it a number of times in this PR (this'll reduce code duplication)

0xbbadbeef reviewed (changes requested):
Things are looking not too bad, just a couple of comments and it should be good to go ✨

causztic commented on lib/constants.js:
```diff
@@ -7,6 +7,7 @@ REOL_CHAN_ID  = '296232884623376384'
 ANNOUNCEMENTS_ID = '413504832356352020'
 PHOTO_GALLERY_ID = '423787873448493056'
 DISCOGRAPHY_ID = '515055814755614730'
+COLORS = ["Red", "Blue", "Green", "Purple", "Yellow", "Orange", "Pink"]
```
The roles we were planning for are actually region based:
```["Scandinavian", "North America", "EU - UK", "EU - Germany", "Asia - SEA", "Asia - Japan", "Australia", "South America"]```
Could you change the roles to these?
I'm not sure how would spaces affect the commands, so you should try those out too

causztic replied:
Sorry, we can leave the regions to another PR

causztic reviewed (changes requested):
Some changes required to get it working for the server!

Commit:
Cleaned up code according to PR comments

Commit:
Delete start.bat

Commit:
Removed experimental regional role info

garrett16r commented:
Everything from your comments except for the helper that @0xbbadbeef mentioned are implemented.  That's one thing that I don't feel like I can do quite right with my current knowledge, so I'll probably come back to that option later (probably when I look at doing regional roles).

causztic commented on lib/util.js:
```diff
@@ -39,4 +39,11 @@ let getChannel = (channel) => {
-module.exports = { catchErrors, setRandomNowPlaying, getChannel }
\ No newline at end of file
+// Checks that the user is sending a command in the #bot-stuff channel
+let inCommandChannel = (msg) => {
```
the reason why I have the `getChannel` method was because I wanted to test it in an admin-only channel when i'm running it locally instead of a public one, so that everyone does not get spammed by my test commands. Can you use `getChannel` as part of `inCommandChannel`? 

Also, `msg.channel.id === BOT_STUFF_ID` already returns a boolean value so you could do something like `return msg.channel.id === BOT_STUFF_ID`

causztic commented on lib/constants.js:
```diff
@@ -7,6 +7,9 @@ REOL_CHAN_ID  = '296232884623376384'
 DISCOGRAPHY_ID = '515055814755614730'
+BOT_STUFF_ID = '432056020060340249'
+COLORS = ["Red", "Blue", "Green"]
+REGIONS = ["NA - US East", "NA - US Central", "NA - US West", "EU - UK", "EU - Germany", "Asia - SEA", "Asia - Japan", "Australia", "South America", "Scandinavian"] // For future use
```
Can you remove the REGIONS constants from this PR?

causztic commented on lib/commands/misc/rmrole.js:
```diff
@@ -0,0 +1,61 @@
+		const memberRoles = msg.member.roles;
+		const channel = this.client.channels.get(getChannel(BOT_STUFF_ID));
+
+		if (!inCommandChannel(msg)) {
```
I think we don't need to return a message if it is not in the correct channel

causztic commented on lib/commands/misc/rmrole.js:
```diff
@@ -0,0 +1,61 @@
+    // Prevent removal of required roles (Can be updated in the future if more required roles are ever added).
+    if (roleArg.toLowerCase() === "member") {
+      msg.reply("You can't remove that role from yourself!");
+      return;
```
code should be consistent! `return msg.reply("You can't remove that role from yourself!");`

causztic commented on lib/commands/misc/giverole.js:
```diff
@@ -25,15 +25,45 @@ module.exports = class GiveRoleCommand extends Command {
+				const memberRole = msg.guild.roles.find(r => r.name === 'Member');
+
+				// Confirms that the user is in the #bot-stuff channel when using the command
+				if (!inCommandChannel(msg)) {
```
I think we don't need to return a message if it is not in the correct channel

Commit:
Removed unneeded code, cleaned up existing

- According to last PR comments, removed unneeded bot responses to using commands in the wrong channel
- Adjusted return statement in rmrole.js to be consistent with other return msg.reply(); statements
- Removed unneeded REGIONS const from constants.js

garrett16r commented:
Forgot to mention in that last commit, I did make sure to implement `getChannel()` in `inCommandChannel()` as well.

causztic reviewed (commented):
One *final* change, could you rename `COLORS` to `AVAILABLE_ROLES`? so that we can extend it to non-color based roles in the future!

Commit:
Renamed COLORS to AVAILABLE_ROLES

causztic reviewed (approved):
Looks good to me!
